### PR TITLE
netdev: Add owl data path sample

### PR DESCRIPTION
**Pull request by dungru**
Support ARP Reply and ICMP packet response

### CODE REVIEW

File: scripts/verify.sh [hunk 1]
```diff
@@ -28,7 +30,7 @@ if [ $final_ret -eq 0 ]; then
 fi
 
 if [ $final_ret -eq 0 ]; then
-    remove_kmod vwifi
+    #remove_kmod vwifi
```
Comment (jserv) [inline scripts/verify.sh:33]: Why did you make such modification?

File: vwifi.c [hunk 2]
```diff
@@ -403,24 +430,163 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+struct net_device_stats *owl_ndo_get_stats(struct net_device *dev)
+{
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    return &owl->stats;
+}
+
+/*
+ * Receive a packet: retrieve, encapsulate and pass over to upper levels
+ */
+void owl_rx(struct net_device *dev)
```
Comment (jserv) [inline vwifi.c:460]: Should be `static`.

File: vwifi.c [hunk 3]
```diff
@@ -257,6 +268,22 @@ static bool is_valid_ssid(const char *connecting_ssid)
...
 }
+
+static void get_connecting_bssid(const char *connecting_ssid,
+                                 u8 *connecting_bssid)
+{
+    struct ap_info_entry_t *ap;
+    struct hlist_node *tmp;
+    u32 key = murmurhash((char *) connecting_ssid);
+    hash_for_each_possible_safe (ssid_table, ap, tmp, node, key) {
+        if (strcmp(connecting_ssid, ap->ssid) == 0) {
```
Comment (jserv) [inline vwifi.c:279]: Rewrite as `if (!strcmp(`

File: vwifi.c [hunk 4]
```diff
@@ -403,24 +430,163 @@ static struct cfg80211_ops owl_cfg_ops = {
     .get_station = owl_get_station,
 };
 
+int owl_ndo_open(struct net_device *dev)
```
Comment (jserv) [inline vwifi.c:433]: All functions starting with `owl_` should be `static`.

File: vwifi.c [hunk 5]
```diff
@@ -403,24 +430,169 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+{
+    /* The first byte is '\0' to avoid being a multicast
+     * address (the first byte of multicast addrs is odd).
+     */
+    memcpy(dev->dev_addr, "\0owl0", ETH_ALEN);
```
Comment (jserv) [inline vwifi.c:438]: How can you ensure that interface is always `owl0`? If we allow multiple `owl` interfaces, `owl1` or `owl2` might appear.

File: vwifi.c [hunk 6]
```diff
@@ -403,24 +430,169 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+    netif_rx(skb);
+    owl->stats.rx_packets++;
+    owl->stats.rx_bytes += pkt->datalen;
+pkt_free:
+    list_del(&pkt->list);
+    kfree(pkt);
+out:
+    return;
+}
+
```
Comment (jserv) [inline vwifi.c:553]: You should keep only one blank line.

File: vwifi.c [hunk 7]
```diff
@@ -403,24 +430,163 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+        // Update Sender
+        u8 tmp[10];
+        memcpy(tmp, &pkt->data[22], 10);
+        ether_addr_copy(&pkt->data[32],
+                        eth->h_source);  // Sender mac is station source mac.
+        memcpy(&pkt->data[22], &pkt->data[32], 10);
+        // Update Target
+        memcpy(&pkt->data[32], tmp, 10);
+    }
+    /* Process ICMP Request */
```
Comment (jserv) [inline vwifi.c:503]: I am not confident if this function is the proper place where ICMP manipulation should locate or not.

Comment (jserv) [reply vwifi.c:503]: If you proceed ICMP requests in this function, where should we implement further TCP/UDP support?

File: vwifi.c [hunk 8]
```diff
@@ -250,13 +261,29 @@ static bool is_valid_ssid(const char *connecting_ssid)
...
+    struct ap_info_entry_t *ap;
+    struct hlist_node *tmp;
+    u32 key = murmurhash((char *) connecting_ssid);
+    hash_for_each_possible_safe (ssid_table, ap, tmp, node, key) {
+        if (strcmp(connecting_ssid, ap->ssid) == 0) {
+            memcpy(connecting_bssid, ap->bssid, ETH_ALEN);
+            break;
+        }
+    }
+    return;
```
Comment (jserv) [inline vwifi.c:284]: Remove `return` statement here.

File: vwifi.c (hunk 7 above)
Comment (dungru) [reply vwifi.c:503]: Doing the protocol stuff in `owl_rx` is not a good method. It's is better to let the kernel network stack handle the protocol. I could try to use  namespace or veth.

Comment (jserv) [reply vwifi.c:503]: Good idea. Look forward to the use of namespace.

File: vwifi.c [hunk 9]
```diff
@@ -403,24 +429,165 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+            memcpy(tmp, &pkt->data[26], 4);
+            memcpy(&pkt->data[26], &pkt->data[30], 4);
+            memcpy(&pkt->data[30], tmp, 4);
+            // icmp reply
+            pkt->data[34] = 0;
+        }
+    }
+    if (!hit_ours_pkt)
+        goto pkt_free;
+    print_hex_dump(KERN_DEBUG, "Sta Rx: ", DUMP_PREFIX_OFFSET, 16, 1, pkt->data,
```
Comment (jserv) [inline vwifi.c:530]: Add a new blank line after `goto pkt_free` statement.

File: vwifi.c [hunk 10]
```diff
@@ -403,24 +429,165 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+    memcpy(&pkt->data, skb->data, skb->len);
+    pkt->datalen = skb->len;
+    list_add_tail(&pkt->list, &owl->rx_queue);  // enqueue
     kfree_skb(skb);
+    /* Directly send to rx_queue, simulate the rx interrupt*/
+    owl_rx(dev);
     return NETDEV_TX_OK;
 }
 
+
```
Comment (jserv) [inline vwifi.c:581]: Don't add this blank line.

File: vwifi.c [hunk 11]
```diff
@@ -403,15 +429,153 @@ static struct cfg80211_ops owl_cfg_ops = {
...
 
+static int owl_ndo_open(struct net_device *dev)
+{
+    /* The first byte is '\0' to avoid being a multicast
+     * address (the first byte of multicast addrs is odd).
+     */
+    char intf_name[ETH_ALEN] = {0};
+    snprintf(&intf_name[1], ETH_ALEN - 1, "%s", dev->name);
+    memcpy(dev->dev_addr, intf_name, ETH_ALEN);
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
```
Comment (jserv) [inline vwifi.c:440]: Move the declaration of `struct owl_context *owl` to the beginning of this function.

File: vwifi.c [hunk 12]
```diff
@@ -250,13 +261,28 @@ static bool is_valid_ssid(const char *connecting_ssid)
...
 }
+
+static void get_connecting_bssid(const char *connecting_ssid,
+                                 u8 *connecting_bssid)
+{
+    struct ap_info_entry_t *ap;
+    struct hlist_node *tmp;
+    u32 key = murmurhash((char *) connecting_ssid);
+    hash_for_each_possible_safe (ssid_table, ap, tmp, node, key) {
+        if (strcmp(connecting_ssid, ap->ssid) == 0) {
```
Comment (jserv) [inline vwifi.c:279]: Rewrite as `if (!strcmp(`.

**Review by jserv — CHANGES_REQUESTED** [review]
Add some descriptions about test procedure in README.md file along with the known issues and limitations.

File: vwifi.c [hunk 13]
```diff
@@ -403,15 +429,155 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+    int offset = 0;
+
+    struct owl_packet *pkt;
+    if (list_empty(&owl->rx_queue)) {
+        printk(KERN_NOTICE "owl rx: No packet in rx_queue\n");
+        goto out;
+    }
+    pkt = list_first_entry(&owl->rx_queue, struct owl_packet, list);
+    /* TODO: Modify the packet and forward to destination */
+    struct ethhdr *eth = (struct ethhdr *) &pkt->data[offset];
```
Comment (jserv) [inline vwifi.c:482]: Don't declare `eth` here.

File: vwifi.c [hunk 14]
```diff
@@ -403,15 +429,155 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+    offset += sizeof(struct ethhdr);
+    /* Process ARP Request */
```
Comment (jserv) [inline vwifi.c:484]: I intend to merge this pull request without major refinement, which decouples protocol stuff from `owl_rx`. However, you shall add informative `FIXME` or `TODO` comments for further involvement.

File: vwifi.c [hunk 15]
```diff
@@ -403,15 +429,155 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+    if (eth->h_proto == htons(ETH_P_ARP)) {
+        printk(KERN_NOTICE "ARP request received, send ARP reply to sta\n");
+        hit_ours_pkt = 1;
+        // dst mac
```
Comment (jserv) [inline vwifi.c:488]: Always use C-style comments. i.e. `/* comments */` rather than `//`.

File: vwifi.c [hunk 16]
```diff
@@ -403,15 +429,155 @@ static struct cfg80211_ops owl_cfg_ops = {
...
+        ether_addr_copy(eth->h_dest, dev->dev_addr);
+        // src mac
+        ether_addr_copy(eth->h_source, owl->connecting_bssid);
+        // arp reply opcode 0x2
```
Comment (jserv) [inline vwifi.c:492]: Explain more about the opcode.

**Comment by dungru** [comment]
Move new request to #15

**Pull request closed without merging**